apt/utils.py

```diff
@@ -0,0 +1,46 @@
+import logging
+from pathlib import Path
+
+from smart_open import smart_open  # type: ignore[import]
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_CHUNK_SIZE = 50 * 1024 * 1024
+
+
+def stream_file_transfer(
+    source_uri: str | Path,
+    target_uri: str | Path,
+    chunk_size: int = DEFAULT_CHUNK_SIZE,
+) -> str:
+    """Stream file from source to target without loading entire file into memory.
+
+    Works with both local paths and remote URIs (e.g., s3://).
+
+    Args:
+        source_uri: Source URI or path to read from
+        target_uri: Target URI or path to write to
+        chunk_size: Size of chunks to read/write in bytes
+    """
+    source_uri = str(source_uri)
+    target_uri = str(target_uri)
+
+    logger.debug(f"Streaming file from '{source_uri}' to '{target_uri}'")
+
+    # ensure parent directory exists for local target paths
+    if "://" not in target_uri and isinstance(target_uri, str):
```

Comment:
Related to previous comment: If `target_uri` was a valid URI for a _local_ file (e.g., `file:///myfolder/file.txt`) was passed (with `myfolder` not an existing directory), is it correct that it would (a) skip the lines ensuring parent directory exists and (b) method will raise an error when `smart_open` tries access a directory that doesn't exist?

Can we update the condition in line 31 to:

```python
from urllib.parse import urlparse

if urllib.parse(target_uri).scheme in ["", "file"] and isinstance(target_uri, str):
...
```

Reply:
Sort of dove-tailing with the longer comment above about URIs, I'm thinking maybe we could keep it a bit more S3 specific with something like:

```python
if not target_uri.startswith("s3://"):
    ...
```

Also don't think we need that second `isinstance()` clause, so removed that too.

Thanks for raising though!  Needed attention either way.